File: Embarcadero/Berlin/CPP/ThingPoint IoT Sample Data Demo/ThingpointCpp/LoggingFrameU.cpp
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------

#include <fmx.h>
#include <EMSHosting.LoggingService.hpp>
#include <System.StrUtils.hpp>

#pragma hdrstop
#include <memory>

#include "LoggingFrameU.h"
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TEMSEdgeLoggingFrame * FFrame;
// ---------------------------------------------------------------------------

struct TThreadHandler : public TCppInterfacedObject<TThreadProcedure> {
    String FCategory;
    TJSONObject * FJSON;

    __fastcall TThreadHandler::TThreadHandler(String ACategory,
	TJSONObject * AJSON) {
	FCategory = ACategory;
	FJSON = AJSON;
    }

    void __fastcall Invoke(void) {
	if (FFrame != NULL)
	    FFrame->LogItem(FCategory, FJSON);
	else
	    delete FJSON;
    }
};

struct TLoggingHandler
    : public TCppInterfacedObject
    <Emshosting::Loggingservice::TEMSLoggingService::TLogProc> {
    void __fastcall Invoke(const System::UnicodeString ACategory,
	System::Json::TJSONObject* const AJSON) {
	TJSONObject * LJSON = dynamic_cast<TJSONObject*>(AJSON->Clone());
	_di_TThreadProcedure handler = new TThreadHandler(ACategory, LJSON);
	TThread::Queue(NULL, handler);
    }
};

__fastcall TEMSEdgeLoggingFrame::TEMSEdgeLoggingFrame(TComponent* Owner)
    : TFrame(Owner) {
    if (FFrame != NULL) {
	throw new Exception("Only one logging frame allowed in an application");
    }
    FFrame = this;
    Emshosting::Loggingservice::TEMSLoggingService::OnLog = NULL;

    Emshosting::Loggingservice::TEMSLoggingService::_di_TLogProc handler =
	new TLoggingHandler();
    Emshosting::Loggingservice::TEMSLoggingService::OnLog = handler;
}

// ---------------------------------------------------------------------------
void __fastcall TEMSEdgeLoggingFrame::ButtonClearClick(TObject *Sender) {
    Memo1->Lines->Clear();
}
// ---------------------------------------------------------------------------

void __fastcall TEMSEdgeLoggingFrame::LogItem(String ACategory,
    TJSONObject * AJSON) {
    std::auto_ptr<TJSONObject>LJSON(new TJSONObject());
    LJSON->AddPair(ACategory, AJSON);
    // Note AJSON is owned by LJSON so AJSON will be freed
    String LLine = LJSON->ToString();
    // Remove line breaks
    LLine = System::Strutils::ReplaceStr(LLine, "\r\n", ", ");
    Memo1->Lines->Add(LLine);
}

File: Embarcadero/Berlin/Object Pascal/ThingPoint ThingConnect IoT Demo/Thingpoint/LoggingFrameU.pas
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
unit LoggingFrameU;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants, 
  FMX.Types, FMX.Graphics, FMX.Controls, FMX.Forms, FMX.Dialogs, FMX.StdCtrls,
  FMX.Layouts, FMX.Controls.Presentation, FMX.ScrollBox, FMX.Memo,
  EMSHosting.ExtensionsServices, System.JSON, IPPeerClient,
  REST.Backend.EMSProvider, REST.Backend.Providers, EMSHosting.EdgeService;

type
  TEMSEdgeLoggingFrame = class(TFrame)
    Memo1: TMemo;
    Layout1: TLayout;
    ButtonClear: TButton;
    CheckBoxLogging: TCheckBox;
    procedure ButtonClearClick(Sender: TObject);
  private
    procedure LogItem(const ACategory: string; const AJSON: TJSONObject);
  protected

  public
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
  end;

implementation

{$R *.fmx}

uses EMSHosting.LoggingService, System.StrUtils;

var
  FFrame: TEMSEdgeLoggingFrame;

procedure TEMSEdgeLoggingFrame.ButtonClearClick(Sender: TObject);
begin
  Memo1.Lines.Clear;
end;

procedure TEMSEdgeLoggingFrame.LogItem(const ACategory: string; const AJSON: TJSONObject);
var
  LJSON: TJSONObject;
  LLine: string;
begin
  LJSON := TJSONObject.Create;
  try
    LJSON.AddPair(ACategory, AJSON); // Note AJSON is owned by LJSON so AJSON will be freed
    LLine := LJSON.ToString;
    // Remove line breaks
    LLine := System.StrUtils.ReplaceStr(LLine, #13#10, ', ');
    Memo1.Lines.Add(LLine);
  finally
    LJSON.Free;
  end;
end;


constructor TEMSEdgeLoggingFrame.Create(AOwner: TComponent);
begin
  inherited;
  if FFrame <> nil then
    raise Exception.Create('Only one logging frame allowed in an application');
  FFrame := Self;

  // Setup logging
  TEMSLoggingService.OnLog :=
    procedure(const ACategory: string; const AJSON: TJSONObject)
    var
      LJSON: TJSONObject;
    begin
      LJSON := AJSON.Clone as TJSONObject;
      TThread.Queue(nil,
        procedure
        begin
          if FFrame <> nil then
            FFrame.LogItem(ACategory, LJSON)
          else
            LJSON.Free;
        end);
    end;
  TEMSLoggingService.OnLoggingEnabled :=
    function: Boolean
    begin
      Result := (FFrame <> nil) and FFrame.CheckBoxLogging.IsChecked;
    end;

end;

destructor TEMSEdgeLoggingFrame.Destroy;
begin
  FFrame := nil;
  inherited;
end;

end.
